src/array/cuda/segment_reduce.cuh

```diff
@@ -155,6 +176,59 @@ void ScatterAdd(
                    n, dim);
 }
 
+/*!
+ * \brief CUDA implementation to update gradients for reduce op max/min
+ * \param graph The input heterogeneous graph.
+ * \param op The binary operator, could be `copy_u`, `copy_e'.
+ * \param list_feat List of the input tensors.
+ * \param list_idx  List of the indices tensors.
+ * \param list_idx_etype List of the node- or edge-type tensors.
+ * \param list_out List of the output tensors.
+ */
+template <typename IdType, typename DType>
+void UpdateGradMinMax_hetero(const HeteroGraphPtr& graph,
+                const std::string& op,
+                const std::vector<NDArray>& list_feat,
+                const std::vector<NDArray>& list_idx,
+                const std::vector<NDArray>& list_idx_types,
+                std::vector<NDArray>* list_out) {
+  if (op == "copy_lhs" || op == "copy_rhs") {
+    std::vector<std::vector<dgl_id_t>> src_dst_ntypes(graph->NumVertexTypes(),
+    std::vector<dgl_id_t>());
+    for (dgl_type_t etype = 0; etype < graph->NumEdgeTypes(); ++etype) {
+      auto pair = graph->meta_graph()->FindEdge(etype);
+      const dgl_id_t dst_ntype = pair.first;  // graph is reversed
+      const dgl_id_t src_ntype = pair.second;
+      auto same_src_dst_ntype = std::find(begin(src_dst_ntypes[dst_ntype]),
+        end(src_dst_ntypes[dst_ntype]), src_ntype);
+      // if op is "copy_lhs", relation type with same src and dst node type will be updated once
+      if (op == "copy_lhs" && same_src_dst_ntype != std::end(src_dst_ntypes[dst_ntype]))
+        continue;
+      src_dst_ntypes[dst_ntype].push_back(src_ntype);
+      const DType* feat_data = list_feat[dst_ntype].Ptr<DType>();
+      const IdType* idx_data = list_idx[dst_ntype].Ptr<IdType>();
+      const IdType* idx_type_data = list_idx_types[dst_ntype].Ptr<IdType>();
+      int type = (op == "copy_lhs") ? src_ntype : etype;
+      DType* out_data = (*list_out)[type].Ptr<DType>();
+      int dim = 1;
+      for (int i = 1; i < (*list_out)[type]->ndim; ++i)
+        dim *= (*list_out)[type]->shape[i];
+      int n = list_feat[dst_ntype]->shape[0];
+      auto *thr_entry = runtime::CUDAThreadEntry::ThreadLocal();
+      const int nbx = FindNumBlocks<'x'>(n);
+      const int ntx = FindNumThreads(dim);
```

Comment:
We may need to handle the case where `dim` < 64 specially, as otherwise we'll get very poor performance.

Reply:
I think the new commit will address this issue too.